Atru/query and import bulk (#942)
hashicorp/web-unified-docs: content/terraform/v1.14.x (alpha)/docs/language/import/bulk.mdx

To generate machine-readable results, you can include the `-json` flag:

```shell-session
$ terraform query -json
```

## Generate configuration for importing results

To generate configuration, run the [`terraform query` command](/terraform/cli/commands/query) and add the `-generate-config-out` flag. The flag expects a path where Terraform stores the generated configuration as a file named `generated.tf`. The `generated.tf` file contains the `resource` and `import` blocks, including resource identities, necessary for importing results.

The following example generates configuration to the working directory:

```shell-session
$ terraform query -generate-config-out=generated.tf
```
 
After generating the results file, you must remove the file before rerunning the command to generate a new results file. Rerunning the command when a `generated.tf` already exists at the specified path results in an error. 

## Import resources

Copy the `import` and `resource` blocks from the `generated.tf` file to your `main.tf` configuration and run the [`terraform apply` command](/terraform/cli/commands/apply) to import the resources.  
  
## Next steps

> After importing your resources, you can discard the query configuration file and the `generated.tf` file, but we recommend keeping them as a historical record.

Reviewer: Copy? Or move? I would assume move otherwise you'd have two couples of these blocks but I'm not sure if there's a special case for the `generated.tf` file

Author: Copying is probably what you want to do. I updated the next steps, though, because it's not clear that we meant that you should apply some process to convert the generated configuration into a historical artifact.

@include 'beta.mdx'

## Introduction

For organizations with large sets of infrastructure resources, manually identifying and importing them is tedious and labor intensive, even when using third-party tools or custom scripts. To alleviate this burden, you can write HCL-based queries and run them with the Terraform CLI to retrieve unmanaged resources so that you can import them in bulk. 

Complete the following steps to find and import resources in bulk:

- Search for resources: Create `list` blocks to search for existing resources.
- Generate configuration: Use the Terraform command to generate `resource` and `import` blocks for importing the resources. The configuration also includes the [resource identity](/terraform/language/import#resource-identity). 
- Import resources: Copy the generated configuration into your main.tf file and run a `terraform apply` command to import the resources.
- After importing resources, you can remove the generated `import` block or keep it as an historical record.

> ## Requirements
> 
> Terraform v1.12 or newer is required to uniquely ID resources according the resource identity. For v1.11 and older, Terraform uses the cloud provider ID attribute. 

Author: ```suggestion
```

To generate machine-readable results, you can include the `-json` flag:

```shell-session
$ terraform query -json
```

## Generate configuration for importing results

To generate configuration, run the [`terraform query` command](/terraform/cli/commands/query) and add the `-generate-config-out` flag. The flag expects a path where Terraform stores the generated configuration as a file named `generated.tf`. The `generated.tf` file contains the `resource` and `import` blocks, including resource identities, necessary for importing results.

The following example generates configuration to the working directory:

```shell-session
$ terraform query -generate-config-out=generated.tf
```
 
After generating the results file, you must remove the file before rerunning the command to generate a new results file. Rerunning the command when a `generated.tf` already exists at the specified path results in an error. 

## Import resources

Copy the `import` and `resource` blocks from the `generated.tf` file to your `main.tf` configuration and run the [`terraform apply` command](/terraform/cli/commands/apply) to import the resources.  
  
## Next steps

> After importing your resources, you can discard the query configuration file and the `generated.tf` file, but we recommend keeping them as a historical record.

Author: ```suggestion
After importing your resources, you can discard the query configuration file and the `generated.tf` file, but we recommend archiving them so that you have a historical record of the import.
```
